new-log-viewer/src/services/decoders/JsonlDecoder/index.ts

```diff
@@ -0,0 +1,242 @@
+import {Nullable} from "../../../typings/common";
+import {
+    Decoder,
+    DecodeResultType,
+    FilteredLogEventMap,
+    JsonlDecoderOptionsType,
+    LogEventCount,
+} from "../../../typings/decoders";
+import {Formatter} from "../../../typings/formatters";
+import {JsonValue} from "../../../typings/js";
+import {
+    INVALID_TIMESTAMP_VALUE,
+    LOG_LEVEL,
+    LogLevelFilter,
+} from "../../../typings/logs";
+import LogbackFormatter from "../../formatters/LogbackFormatter";
+import {
+    convertToDayjsTimestamp,
+    convertToLogLevelValue,
+    isJsonObject,
+    JsonLogEvent,
+} from "./utils";
+
+
+/**
+ * A decoder for JSONL (JSON lines) files that contain log events. See `JsonlDecoderOptionsType` for
+ * properties that are specific to json log events (compared to generic JSON records).
+ */
+class JsonlDecoder implements Decoder {
+    static #textDecoder = new TextDecoder();
+
+    #dataArray: Nullable<Uint8Array>;
+
+    #logLevelKey: string;
+
+    #timestampKey: string;
+
+    #logEvents: JsonLogEvent[] = [];
+
+    #filteredLogEventMap: FilteredLogEventMap = null;
+
+    #invalidLogEventIdxToRawLine: Map<number, string> = new Map();
+
+    #formatter: Formatter;
+
+    /**
+     * @param dataArray
+     * @param decoderOptions
+     */
+    constructor (dataArray: Uint8Array, decoderOptions: JsonlDecoderOptionsType) {
+        this.#dataArray = dataArray;
+        this.#logLevelKey = decoderOptions.logLevelKey;
+        this.#timestampKey = decoderOptions.timestampKey;
+        this.#formatter = new LogbackFormatter({formatString: decoderOptions.formatString});
+    }
+
+    getEstimatedNumEvents (): number {
+        return this.#logEvents.length;
+    }
+
+    getFilteredLogEventMap (): FilteredLogEventMap {
+        return this.#filteredLogEventMap;
+    }
+
+    setLogLevelFilter (logLevelFilter: LogLevelFilter): boolean {
+        this.#filterLogs(logLevelFilter);
+
+        return true;
+    }
+
+    build (): LogEventCount {
+        this.#deserialize();
+
+        const numInvalidEvents = Array.from(this.#invalidLogEventIdxToRawLine.keys()).length;
+
+        return {
+            numValidEvents: this.#logEvents.length - numInvalidEvents,
+            numInvalidEvents: numInvalidEvents,
+        };
+    }
+
+    setFormatterOptions (options: JsonlDecoderOptionsType): boolean {
+        this.#formatter = new LogbackFormatter({formatString: options.formatString});
+
+        return true;
+    }
+
+    decodeRange (
+        beginIdx: number,
+        endIdx: number,
+        useFilter: boolean,
+    ): Nullable<DecodeResultType[]> {
+        if (useFilter && null === this.#filteredLogEventMap) {
+            return null;
+        }
+
+        // Prevents typescript potential null warning.
+        const filteredLogEventIndices: number[] = this.#filteredLogEventMap as number[];
+
+        const length: number = useFilter ?
+            filteredLogEventIndices.length :
+            this.#logEvents.length;
```

Comment:
```suggestion
        const length: number = null !== this.#filteredLogEventMap ?
            this.#filteredLogEventMap.length :
            this.#logEvents.length;
```

Since we validate `this.#filteredLogEventMap` is only non-null when `useFilter` is set, then we can use `null !== this.#filteredLogEventMap` in place of `useFilter` and get rid of some of the explicit casts.

Reply:
Even with `FilteredLogEventMap` being exported from decoders.ts?